### PR TITLE
Add/Edit User pages (Tenants & Users section)

### PR DESCRIPTION
**Pull request by xantorres**
Also fixes some issues and errors in ValidatedForm and validations

TODO: dev server api resources and action create/update functions

image

### CODE REVIEW

File: src/app/core/utils/fieldValidators.js [hunk 1]
```diff
@@ -49,32 +54,50 @@ export const lengthValidator = (minLength, maxLength) =>
...
+    hasOneUpperChar,
+    hasOneNumber,
+    hasOneSpecialChar,
+  ])(value),
+  cond([
+    [complement(hasMinLength(8)), always('Password must be at least 8 characters long')],
+    [complement(hasOneLowerChar), always('Password must contain at least one lowercase letter')],
+    [complement(hasOneUpperChar), always('Password must contain at least one uppercase letter')],
+    [complement(hasOneNumber), always('Password must contain at least one number')],
+    [complement(hasOneSpecialChar), always('Password must contain at least one special character')],
```
Comment (brennancheung) [inline src/app/core/utils/fieldValidators.js]: Might be able to write a small function to simplify some of this repetition and remove the `complement` and `always` that add a little clutter.  Also, the list is being repeated twice so probably only need to define it once.

File: src/app/core/utils/fieldValidators.js [hunk 2]
```diff
@@ -20,27 +20,32 @@ export const customValidator = (validator, errorMessage) =>
   new FieldValidator(validator, errorMessage)
 
 const fieldIsUnset = value => isNil(value) || isEmpty(value) || value === false
+export const hasMinLength = moize(minLen => both(is(String), val => val.length >= minLen))
```
Comment (brennancheung) [inline src/app/core/utils/fieldValidators.js]: Do we **really** need to use `moize` here:interrobang:

File: src/app/core/utils/fieldValidators.js (hunk 1 above)
Comment (xantorres) [reply src/app/core/utils/fieldValidators.js]: The second time is a switch to show a different message depending on the error

File: src/app/core/utils/fieldValidators.js (hunk 2 above)
Comment (xantorres) [reply src/app/core/utils/fieldValidators.js]: It doesn't hurt

Comment (brennancheung) [reply src/app/core/utils/fieldValidators.js]: Creating this function is cheap.  Probably faster than the time it takes to do the extra memory lookup from memoization.  Modern CPU's can execute anywhere from 50-200 instructions in the time it takes to do a single memory fetch operation depending on if it is in L1/L2/main memory/TLB miss.  Even if it was slower to create than perform the lookup, we don't need to optimize things that are going to have less than a nanosecond of difference to the end user.

http://wiki.c2.com/?PrematureOptimization

Comment (xantorres) [reply src/app/core/utils/fieldValidators.js]: as I always say, this is not about creating the function cost, is about passing new props to children components that causes React engine to re-render, and that always add some overhead, specially in large forms with a lot of inputs and sub-components, so micro optimizations and memoization techniques like this are the way to go

Comment (xantorres) [reply src/app/core/utils/fieldValidators.js]: anyway, I have checked, and it seems I'm not passing down the result of the memoized function as a prop (I'm using it to check if the password had a minimal length), so it is not be needed at all in this case.

**Pull request merged**